File: app/api/chat/stream/route.js
import { NextResponse } from 'next/server';
import { metrics_service } from '@/lib/monitoring/metrics_service';
import { auth } from '@/auth';
import { createLangSmithClient, LangChainTracer } from 'langsmith';
import { OpenAI } from 'openai';
import { v4 as uuidv4 } from 'uuid';
import { createClient } from '@supabase/supabase-js';

/**
 * Streaming Chat API Endpoint
 * Provides real-time token-by-token chat responses with
 * integrated LangSmith tracing and metrics tracking.
 * Location: /app/api/chat/stream/route.js
 */

// Initialize OpenAI client
const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY
});

// Initialize LangSmith client
const langsmith = createLangSmithClient({
  apiKey: process.env.LANGSMITH_API_KEY,
  projectName: 'climate-economy-chat'
});

// Initialize Supabase client
const supabaseUrl = process.env.SUPABASE_URL;
const supabaseKey = process.env.SUPABASE_SERVICE_KEY;
const supabase = createClient(supabaseUrl, supabaseKey);

// System prompt
const SYSTEM_PROMPT = `You are a specialized assistant for the Massachusetts Clean Tech Ecosystem. Your purpose is to help individuals in Massachusetts find jobs, training, and resources in the clean energy economy.

Special considerations:
1. For Veterans: Help translate military experience to clean energy careers
2. For International Professionals: Help evaluate overseas credentials for Massachusetts 
3. For Environmental Justice Communities: Prioritize opportunities in Gateway Cities

Focus on Massachusetts-specific information whenever possible. If you don't have specific Massachusetts information, clearly indicate this.

Your tone should be helpful, informative, and encouraging.

Format your responses using markdown for better readability. Use bullet points, headers, and other formatting when appropriate.
`;

export async function POST(request) {
  // Create a unique run ID for tracing
  const runId = uuidv4();
  const tracer = new LangChainTracer({
    projectName: 'climate-economy-chat',
    client: langsmith,
    run_id: runId,
    name: 'streaming_chat'
  });
  
  // Start timing
  const startTime = Date.now();
  
  try {
    // Track this as a trace
    await tracer.startTrace({
      name: 'streaming_chat_request',
      input: { 
        headers: request.headers,
        method: request.method
      }
    });
    
    // Get user session for authentication
    const session = await auth();
    const userId = session?.user?.id || 'anonymous';
    
    // Extract query and model from request
    const body = await request.json();
    const { query, model = 'gpt-4o' } = body;
    
    if (!query || query.trim() === '') {
      await tracer.endTrace({
        output: { error: 'Missing query parameter' },
        error: new Error('Missing query parameter')
      });
      return NextResponse.json(
        { message: 'Query parameter is required' }, 
        { status: 400 }
      );
    }
    
    // Track search event
    await metrics_service.track_event(
      'streaming_chat',
      userId,
      {
        query,
        model,
        is_streaming: true
      }
    );
    
    // Get relevant context for the query
    const contextStep = await tracer.trackStep('retrieve_context', {
      input: { query }
    });
    
    const context = await getRelevantContext(query, tracer);
    
    await tracer.endStep(contextStep, {
      output: { 
        contextCount: context.length,
        sources: context.map(c => c.source)
      }
    });
    
    // Format context for the prompt
    const contextText = context
      .map(item => `Source: ${item.source}\n${item.content}`)
      .join('\n\n');
    
    // Create messages array for the chat completion
    const messages = [
      { role: 'system', content: SYSTEM_PROMPT },
      { role: 'user', content: `Massachusetts Clean Energy Query: ${query}\n\nRelevant Context:\n${contextText}` }
    ];
    
    // Create a stream from OpenAI
    const streamStep = await tracer.trackStep('generate_stream', {
      input: { 
        model,
        messages: messages.map(m => ({ role: m.role, content: m.content.substring(0, 100) + '...' }))
      }
    });
    
    const stream = await openai.chat.completions.create({
      model,
      messages,
      stream: true,
      temperature: 0.7,
      max_tokens: 1500
    });
    
    // Create a readable stream to return to the client
    const encoder = new TextEncoder();
    let responseText = '';
    let tokenCount = 0;
    
    const readableStream = new ReadableStream({
      async start(controller) {
        // Process each chunk from OpenAI
        for await (const chunk of stream) {
          if (chunk.choices[0]?.delta?.content) {
            const content = chunk.choices[0].delta.content;
            responseText += content;
            tokenCount++;
            
            // Send the chunk to the client
            controller.enqueue(encoder.encode(content));
          }
        }
        
        // Send the sources at the end as JSON
        const sources = context.map(item => ({
          source: item.source,
          url: item.metadata?.url || '',
          relevance: item.relevance_score || 0
        }));
        
        controller.enqueue(encoder.encode(`\n\nSOURCES:${JSON.stringify(sources)}`));
        controller.close();
        
        // End the streaming step
        await tracer.endStep(streamStep, {
          output: { 
            tokenCount,
            responseLength: responseText.length
          }
        });
        
        // Calculate performance metrics
        const endTime = Date.now();
        const duration = endTime - startTime;
        
        // Track API performance
        await metrics_service.track_api_performance(
          '/api/chat/stream',
          duration,
          200,
          userId,
          { query, model }
        );
        
        // Store the chat in memory (asynchronously, don't wait)
        storeChat(userId, query, responseText);
        
        // End trace with results
        await tracer.endTrace({
          output: {
            responseLength: responseText.length,
            tokenCount,
            duration_ms: duration,
            context: {
              count: context.length,
              sources: context.map(c => c.source)
            }
          }
        });
      }
    });
    
    // Return the stream
    return new Response(readableStream, {
      headers: {
        'Content-Type': 'text/event-stream',
        'Cache-Control': 'no-cache',
        'Connection': 'keep-alive',
      }
    });
  } catch (error) {
    console.error('Streaming chat error:', error);
    
    // End tracing with error
    await tracer.endTrace({
      output: { error: error.message },
      error
    });
    
    // Return error response
    return NextResponse.json(
      { message: 'Error generating response', error: error.message },
      { status: 500 }
    );
  }
}

/**
 * Get relevant context for the query from multiple sources
 */
async function getRelevantContext(query, tracer) {
  const context = [];
  
  try {
    // Get database results first
    const dbStep = await tracer.trackStep('database_retrieval', {
      input: { query }
    });
    
    try {
      // Calculate embedding for vector search
      const { data: embedding } = await supabase.functions.invoke('embed-text', {
        body: { text: query }
      });
      
      if (embedding?.vector) {
        // Perform vector search
        const { data: matches, error } = await supabase
          .from('climate_memories')
          .select('*')
          .order('created_at', { ascending: false })
          .limit(5)
          .order('similarity', { ascending: false })
          .rpc('match_climate_memories', {
            query_embedding: embedding.vector,
            match_threshold: 0.6,
            match_count: 5
          });
        
        if (matches && !error) {
          // Add database results to context
          for (const match of matches) {
            context.push({
              source: 'database',
              content: match.content,
              metadata: match.metadata || {},
              relevance_score: match.similarity,
              category: match.category
            });
          }
        }
      }
      
      await tracer.endStep(dbStep, {
        output: { resultCount: context.length }
      });
    } catch (dbError) {
      console.error('Database search error:', dbError);
      await tracer.endStep(dbStep, {
        output: { error: dbError.message },
        error: dbError
      });
    }
    
    // If we don't have enough database results, supplement with web search
    if (context.length < 3) {
      const webStep = await tracer.trackStep('web_search', {
        input: { query }
      });
      
      try {
        // Append Massachusetts if not already in query
        let searchQuery = query;
        if (!query.toLowerCase().includes('massachusetts') && 
            !query.toLowerCase().includes('mass') && 
            !query.toLowerCase().includes('ma')) {
          searchQuery = `${query} Massachusetts`;
        }
        
        // Call Serper API for web search
        const response = await fetch('https://google.serper.dev/search', {
          method: 'POST',
          headers: {
            'X-API-KEY': process.env.SERPER_API_KEY,
            'Content-Type': 'application/json'
          },
          body: JSON.stringify({
            q: searchQuery,
            num: 5
          })
        });
        
        if (response.ok) {
          const data = await response.json();
          
          // Add web results to context
          if (data.organic && data.organic.length > 0) {
            data.organic.forEach((item, index) => {
              // Calculate a score (decreasing by position)
              const positionScore = 1 - (index * 0.1);
              
              context.push({
                source: 'web',
                content: `Title: ${item.title}\nSnippet: ${item.snippet}\nURL: ${item.link}`,
                metadata: { url: item.link },
                relevance_score: positionScore
              });
            });
          }
        }
        
        await tracer.endStep(webStep, {
          output: { 
            resultCount: context.length - (context.filter(c => c.source === 'database').length)
          }
        });
      } catch (webError) {
        console.error('Web search error:', webError);
        await tracer.endStep(webStep, {
          output: { error: webError.message },
          error: webError
        });
      }
    }
    
    // Sort by relevance score
    context.sort((a, b) => (b.relevance_score || 0) - (a.relevance_score || 0));
    
    // Limit to top 5 most relevant items
    return context.slice(0, 5);
  } catch (error) {
    console.error('Error getting relevant context:', error);
    return [];
  }
}

/**
 * Store the chat in memory for later retrieval
 */
async function storeChat(userId, query, response) {
  try {
    // Store in Supabase
    await supabase
      .from('climate_memories')
      .insert([
        {
          user_id: userId,
          content: `Q: ${query}\nA: ${response}`,
          category: 'conversation',
          source: 'assistant',
          metadata: { 
            is_conversation: true,
            timestamp: new Date().toISOString()
          }
        }
      ]);
  } catch (error) {
    console.error('Error storing chat:', error);
  }
} 

File: app/api/climate-chat/route.py
from fastapi import FastAPI, HTTPException, Request
from fastapi.responses import StreamingResponse
import json
import os
import time
import openai
import uuid
from typing import Dict, Any, List, Optional
from dotenv import load_dotenv
import asyncio

# Import consolidated services
from lib.memory.memory_service import MemoryService, ClimateMemoryEntry, MemoryServiceError
from lib.redis.redis_service import RedisService, RedisServiceError
from lib.tracing.langsmith_service import TracingService, TracingServiceError

# Import tools
from tools.base_tool import BaseTool, ToolError
from tools.ej_geospatial import EJGeospatialTool
from tools.international_credential_evaluator import InternationalCredentialEvaluator

# Load environment variables
load_dotenv()

# Initialize services
memory_service = MemoryService()
redis_service = RedisService()
tracing_service = TracingService()

# Initialize tools
ej_geospatial_tool = EJGeospatialTool()
international_credential_evaluator = InternationalCredentialEvaluator()

# Initialize OpenAI client
openai_client = openai.OpenAI(api_key=os.environ.get("OPENAI_API_KEY"))

app = FastAPI()

SYSTEM_PROMPT = """You are a specialized assistant for the Massachusetts Clean Tech Ecosystem. Your purpose is to help individuals in Massachusetts find jobs, training, and resources in the clean energy economy.

Special considerations:
1. For Veterans: Help translate military experience to clean energy careers
2. For International Professionals: Help evaluate overseas credentials for Massachusetts
3. For Environmental Justice Communities: Prioritize opportunities in Gateway Cities

Focus on Massachusetts-specific information whenever possible. If you don't have specific Massachusetts information, clearly indicate this.

Your tone should be helpful, informative, and encouraging.
"""

async def get_relevant_context(query: str, user_id: str, run_id: Optional[str] = None) -> List[Dict[str, Any]]:
    """Get relevant context from memory and web search for the climate chat"""
    try:
        # Create a trace for context retrieval
        with tracing_service.trace(
            name="get_relevant_context",
            run_id=run_id or tracing_service.create_run_id(),
            inputs={"query": query, "user_id": user_id}
        ) as trace_run:
            # Try to get cached context first
            try:
                cache_key = f"context:{user_id}:{query}"
                cached_context = await redis_service.get(cache_key)
                if cached_context:
                    trace_run.end(outputs={"source": "cache", "context_count": len(cached_context)})
                    return cached_context
            except RedisServiceError:
                # Continue if Redis is unavailable
                pass
            
            context = []
            
            # First, try to get information from memory
            try:
                # Search user's memories
                memory_results = await memory_service.search_memories(
                    query=query,
                    user_id=user_id,
                    limit=5,
                    categories=["report", "job", "training"]
                )
                
                # Add memory results to context
                for result in memory_results:
                    context.append({
                        "source": "memory",
                        "content": result["content"],
                        "metadata": result["metadata"],
                        "relevance_score": result.get("relevance_score", 0.0)
                    })
                    
            except MemoryServiceError as e:
                # Log the error but continue
                print(f"Error searching memories: {str(e)}")
            
            # If we have EJ community data, add it to the context
            try:
                # Get user profile
                user_profile = await memory_service.get_user_profile(user_id)
                
                if user_profile and user_profile.is_ej_community and user_profile.location:
                    # Get EJ community data
                    ej_data = await ej_geospatial_tool.run(
                        action="get_community_profile",
                        city=user_profile.location
                    )
                    
                    if ej_data:
                        context.append({
                            "source": "ej_data",
                            "content": f"The user is from {user_profile.location}, which is an Environmental Justice community with the following characteristics: {json.dumps(ej_data)}",
                            "metadata": {"type": "ej_community_data"},
                            "relevance_score": 0.9  # High relevance for personalized data
                        })
            except Exception as e:
                # Log the error but continue
                print(f"Error getting EJ data: {str(e)}")
            
            # If we have international professional data, add it to the context
            try:
                if user_profile and user_profile.has_international_credentials and user_profile.origin_country:
                    # Get international credential data
                    credential_data = user_profile.international_credentials
                    
                    if credential_data:
                        context.append({
                            "source": "international_data",
                            "content": f"The user is an international professional from {user_profile.origin_country} with credentials that may need evaluation: {json.dumps(credential_data)}",
                            "metadata": {"type": "international_credential_data"},
                            "relevance_score": 0.9  # High relevance for personalized data
                        })
            except Exception as e:
                # Log the error but continue
                print(f"Error getting international data: {str(e)}")
            
            # Sort by relevance score
            context.sort(key=lambda x: x.get("relevance_score", 0), reverse=True)
            
            # Limit to top 5 most relevant items
            context = context[:5]
            
            # Cache the context in Redis if available
            try:
                cache_key = f"context:{user_id}:{query}"
                await redis_service.set(cache_key, context, 3600)  # Cache for 1 hour
            except RedisServiceError:
                # Continue if Redis is unavailable
                pass
            
            # End the trace with the final context
            trace_run.end(outputs={
                "context_count": len(context),
                "sources": [item["source"] for item in context]
            })
            
            return context
    except Exception as e:
        print(f"Error getting relevant context: {str(e)}")
        # If there's an error, return an empty context
        return []

async def chat_completion_with_streaming(query: str, user_id: str, stream_tokens: bool = False):
    """Generate a chat completion with optional streaming"""
    # Create a run ID for tracing
    run_id = tracing_service.create_run_id()
    
    try:
        # Create a trace for the chat completion
        with tracing_service.trace(
            name="climate_chat_completion",
            run_id=run_id,
            inputs={"query": query, "user_id": user_id, "stream_tokens": stream_tokens}
        ) as trace_run:
            start_time = time.time()
            
            # Get relevant context
            context = await get_relevant_context(query, user_id, run_id)
            
            # Format context for the prompt
            context_text = ""
            for item in context:
                source = item["source"]
                content = item["content"]
                context_text += f"[{source}]\n{content}\n\n"
            
            # Create messages for the chat completion
            messages = [
                {"role": "system", "content": SYSTEM_PROMPT},
                {"role": "system", "content": f"Context information:\n{context_text}"},
                {"role": "user", "content": query}
            ]
            
            # Generate completion
            if stream_tokens:
                # Stream the response
                completion_stream = openai_client.chat.completions.create(
                    model="gpt-4",
                    messages=messages,
                    temperature=0.7,
                    stream=True
                )
                
                # Define the generator function for streaming
                async def generate():
                    try:
                        # Start the response
                        yield "data: {\"type\": \"start\", \"run_id\": \"" + run_id + "\"}\n\n"
                        
                        # Stream the tokens
                        full_response = ""
                        for chunk in completion_stream:
                            if chunk.choices and chunk.choices[0].delta.content:
                                content = chunk.choices[0].delta.content
                                full_response += content
                                yield f"data: {json.dumps({'type': 'token', 'content': content})}\n\n"
                        
                        # End the response
                        end_time = time.time()
                        duration_ms = int((end_time - start_time) * 1000)
                        
                        # Store the response in memory
                        try:
                            memory_entry = ClimateMemoryEntry(
                                content=full_response,
                                user_id=user_id,
                                category="chat",
                                metadata={
                                    "query": query,
                                    "timestamp": time.time(),
                                    "context_sources": [item["source"] for item in context]
                                }
                            )
                            await memory_service.add_memory(memory_entry)
                        except MemoryServiceError as e:
                            print(f"Error storing chat in memory: {str(e)}")
                        
                        # End the trace
                        trace_run.end(outputs={
                            "response_length": len(full_response),
                            "response_time_ms": duration_ms
                        })
                        
                        yield f"data: {json.dumps({'type': 'end', 'duration_ms': duration_ms})}\n\n"
                    except Exception as e:
                        error_msg = str(e)
                        print(f"Error in streaming: {error_msg}")
                        yield f"data: {json.dumps({'type': 'error', 'error': error_msg})}\n\n"
                
                return generate()
            else:
                # Generate the full response at once
                completion = openai_client.chat.completions.create(
                    model="gpt-4",
                    messages=messages,
                    temperature=0.7
                )
                
                response = completion.choices[0].message.content
                
                # Store the response in memory
                try:
                    memory_entry = ClimateMemoryEntry(
                        content=response,
                        user_id=user_id,
                        category="chat",
                        metadata={
                            "query": query,
                            "timestamp": time.time(),
                            "context_sources": [item["source"] for item in context]
                        }
                    )
                    await memory_service.add_memory(memory_entry)
                except MemoryServiceError as e:
                    print(f"Error storing chat in memory: {str(e)}")
                
                # End the trace
                end_time = time.time()
                duration_ms = int((end_time - start_time) * 1000)
                trace_run.end(outputs={
                    "response_length": len(response),
                    "response_time_ms": duration_ms
                })
                
                return {
                    "response": response,
                    "run_id": run_id,
                    "duration_ms": duration_ms
                }
    except Exception as e:
        error_msg = str(e)
        print(f"Error in chat completion: {error_msg}")
        return {
            "error": "Error generating response",
            "message": error_msg
        }

@app.post("/api/climate-chat")
async def climate_chat(request: Request):
    """API endpoint for climate chat"""
    try:
        # Parse request
        data = await request.json()
        query = data.get("query")
        user_id = data.get("user_id")
        stream_tokens = data.get("stream_tokens", False)
        
        # Validate request
        if not query:
            raise HTTPException(status_code=400, detail="Query is required")
        if not user_id:
            raise HTTPException(status_code=400, detail="User ID is required")
        
        # Generate response
        if stream_tokens:
            # Return streaming response
            return StreamingResponse(
                chat_completion_with_streaming(query, user_id, True),
                media_type="text/event-stream"
            )
        else:
            # Return regular response
            return await chat_completion_with_streaming(query, user_id, False)
    except Exception as e:
        # Handle errors
        error_msg = str(e)
        print(f"Error in climate chat endpoint: {error_msg}")
        raise HTTPException(status_code=500, detail=error_msg)
